api/src/main/java/org/apache/iceberg/expressions/VariantExpressionUtil.java

```diff
@@ -111,8 +113,19 @@ static <T> T castTo(VariantValue value, Type type) {
         }
 
         break;
-    }
+      case TIMESTAMP:
+      case TIMESTAMP_NANO:
+      case TIME:
```

Comment:
Currently, we are supporting INT64 to those types. How about other conversions from INT8/INT16/INT32? Also how about TIMESTAMP =>  TIMESTAMP_NANO and also TIME => TIMESTAMP / TIMESTAMP_NANO? 

cc @RussellSpitzer and @rdblue  I think we need to handle such conversions?

Reply:
I don't think that any conversion other than when types match should be allowed. This is equivalent to `Literal#to(Type)` and we don't allow conversion from date/time to int32 or int64 there. Why would this be necessary?

Reply:
I see. It's to fix some common issues: so seems TIMESTAMP => TIMESTAMP_NANO should be supported. Is TIME => TIMESTAMP / TIMESTAMP_NANO allowed - maybe not?

>    * <p>When a predicate is bound to a concrete data column, literals are converted to match the
>    * bound column's type. This conversion process is more narrow than a cast and is only intended
>    * for cases where substituting one type is a common mistake (e.g. 34 instead of 34L) or where
>    * this API avoids requiring a concrete class (e.g., dates).

Reply:
I had the same question too while working on this. But then, `TimeLiteral` did not convert its value to TIMESTAMP/TIMESTAMP_NANO and vice-versa, hence did not carried out such conversion in `VariantExpressionUtil`